1 add test suite

emjbishop opened the pull request:
This PR adds a test suite and automatic checks.

**Changes**

- New test suite:
    - Added `tests/` folder containing tests for main scripts.
    - Updated example `fasta_dir` files to allow for more robust testing.

- New GitHub Actions:
    - Added `R-CMD-check.yaml` to run R CMD check.
    - Added `test-coverage.yaml` to check test coverage with `covr`.

- Documentation updates:
    - Added examples to function documentation.
    - Add badge to README for R CMD check results.

- Use the built-in `message()`, `warning()`, and `error()` functions for logging throughout instead of custom functions.

**Bug fixes**

- Fixed logic in `build_lookup_from_fastas()` so that Adaptive lookup tables are created properly, and in a way more similar to the Python version ([TCRconvert](https://github.com/seshadrilab/tcrconvert)).
- Fixed `convert_gene()` to preserve input row order after `merge()`.

sckott reviewed (changes requested):
Nice work! looking great, some comments and requests for change

- I'd consider using https://styler.r-lib.org/ package to make the syntax/style consistent throughout the package
- have a look at the `goodpractice` output, ropensci uses it for their reviews, gives some good tips, run `goodpractice::gp()` after installing it - it's recommending to try to have line lengths no longer than 80 characters, avoid using `sapply` (consider using `vapply`), avoid expressions like `1:nrow(df)`
- I worry about so much information contained in code comments. Comments are often forgotten about and tend to get out of date. I'd suggest to move important comments within function code into the roxygen2 lines above the function using appropriate roxygen2 tags (e.g., details, section, etc.)

emjbishop commented on R/convert.R:
```diff
@@ -104,20 +106,27 @@ convert_gene <- function(df, frm, to, species = "human", frm_cols = NULL, quiet
+  # Loop over gene columns, doing a merge to get converted gene names
   new_genes <- list()
-  bad_genes <- c()
+  bad_genes_all <- c()
```
Good to know, I wish could pre-populate but in this case I don't know how many NA's I'm going to end up with ahead of time.

emjbishop commented on R/convert.R:
```diff
@@ -18,17 +14,19 @@ col_ref <- list(
   if (frm == "tenx") {
     lookup_f <- file.path(data_path, "lookup_from_tenx.csv")
-    log_message("WARNING", "Converting from 10X which lacks allele info. Choosing *01 as allele for all genes.")
+    message("Converting from 10X which lacks allele info. Choosing *01 as allele for all genes.")
```
Based on feedback from the lab users, I made some messages warnings so they'll always appear. Based on our conversation I opted to use a 'verbose' strategy instead of 'quiet'. Fixed in commit 87eab8a8570ee73c4c2ceafa826e8989c5547445.

emjbishop commented on R/convert.R:
```diff
@@ -81,20 +81,22 @@ which_frm_cols <- function(df, frm, frm_cols = NULL) {
-    stop("ValueError")
+    stop('"frm" and "to" formats should be different.')
   }
-  if (nrow(df) == 0) {
```
Good idea, fixed in cb590bd95ae4fa0f0d7d38709e3d59793192ea55

emjbishop commented on R/convert.R:
```diff
@@ -81,20 +81,22 @@ which_frm_cols <- function(df, frm, frm_cols = NULL) {
-  if (nrow(df) == 0) {
-    log_message("ERROR", "Input data is empty.")
-    stop("ValueError")
+  if (nrow(df) < 1 | is.null(nrow(df))) {
```
I didn't know that, thanks for pointing this out. However this is no longer needed due to changes made in cb590bd95ae4fa0f0d7d38709e3d59793192ea55

emjbishop commented on R/convert.R:
```diff
@@ -126,6 +135,12 @@ convert_gene <- function(df, frm, to, species = "human", frm_cols = NULL, quiet
+  df_out <- subset(df_out, select = -c(id))
+
+  # Replace NoData with NA
+  df_out[df_out == "NoData"] <- NA
```
Good idea, fixed in 36ca327b5742764d520cf792413222b85ae1a5ee.

emjbishop commented on R/build_lookup.R:
```diff
@@ -39,23 +58,29 @@ extract_imgt_genes <- function(data_dir) {
+#' @examples
+#' TCRconvertR:::add_dash_one('TRBV2*01')
 add_dash_one <- function(gene_str) {
   if (!grepl("-", gene_str)) {
```
Added a line that will stop with error if length > 1 in f751fdfa1f8d14a8b5cd2ea3f41ab5b97084ef20.

emjbishop commented on R/convert.R:
```diff
@@ -18,17 +14,19 @@ col_ref <- list(
 #'
 #' @return A string, the path to correct lookup table.
+#' @examples
+#' TCRconvertR:::choose_lookup("imgt", "adaptive")
```
I used the triple colons because `run_examples()` didn't work without them. However, since you told me about using `@export` with `@keywords internal` I've switched to that. Fixed in 7a762a084fe846917d4a17ceb2ac37393c4771fa.

emjbishop commented on R/convert.R:
```diff
@@ -50,19 +47,22 @@ choose_lookup <- function(frm, to, species = "human") {
+#' @examples
+#' tcr_file <- get_example_path('tenx.csv')
+#' df <- read.csv(tcr_file)
+#' TCRconvertR:::which_frm_cols(df, 'tenx')
```
Fixed in 7a762a084fe846917d4a17ceb2ac37393c4771fa.

emjbishop commented on tests/testthat/fixtures/fasta_dir/lookup_from_tenx.csv:
You're right, the files are almost the same and the tests only use the one in `examples/`. I think `fixtures/` is a remnant from me learning how to build tests in R. I've removed it in 4b2cd0b2a67e8624edf6fa0b2cd95a5612aa1e71.

sckott commented on R/build_lookup.R:
```diff
@@ -39,23 +62,36 @@ extract_imgt_genes <- function(data_dir) {
+#' @examples
+#' add_dash_one('TRBV2*01')
 add_dash_one <- function(gene_str) {
+  if (!length(gene_str) == 1) {
```
beware with differences between R and python for string length. `length("abc") == 1`  in R and `len("abc") == 3` in Python

sckott replied:
if you want the number of characters in a string, do `nchar()`. 

length does that in R i think because R is vectorized whenever possible, so length of a vector of strings is just the number of strings in the vector `length(letters) == 26`

sckott replied:
maybe you do want to make sure it's only 1 string in which case what you have is great

emjbishop replied:
Yes I do just want to check it's one string in this case, but I've been burned by the `length()` `len()` difference recently so thanks for bringing it up!

sckott commented on R/build_lookup.R:
```diff
@@ -3,13 +3,22 @@
   lines <- readLines(infile)
   
   # Extract gene names from headers
   imgt_list <- sapply(lines[grep("^>", lines)], function(line) {
```
I would suggest avoid `sapply` and using `vapply` or something else

emjbishop replied:
Fixed in 64b0f82d072005247fc31373c9bdbdb59ed4619f

sckott commented on R/convert.R:
```diff
@@ -16,19 +12,27 @@ col_ref <- list(
+#' @keywords internal
+#' @examples
+#' choose_lookup("imgt", "adaptive")
+choose_lookup <- function(frm, to, species = "human", verbose) {
```
I think `verbose` param should have a default value, yes? if not, then it should be placed to the left of the parameters WITH default values, with `frm` and `to`

emjbishop replied:
Yes, this was an oversight on my part. I fixed this and added tests in e968cf790898e678da8d58d9a37837b257bb09a1.

sckott commented on R/convert.R:
```diff
@@ -48,21 +51,29 @@ choose_lookup <- function(frm, to, species = "human") {
+#' tcr_file <- get_example_path('tenx.csv')
+#' df <- read.csv(tcr_file)
+#' which_frm_cols(df, 'tenx')
+which_frm_cols <- function(df, frm, frm_cols = NULL, verbose) {
```
same here about default value and placement

emjbishop replied:
Fixed in e968cf790898e678da8d58d9a37837b257bb09a1.

sckott commented on R/convert.R:
```diff
@@ -126,6 +149,12 @@ convert_gene <- function(df, frm, to, species = "human", frm_cols = NULL, quiet
+  # Replace NoData with NA
+  df_out[df_out == "NoData"] <- NA_character_
+
   return(df_out)
```
you don't need to explicitly use `return` as you do in python, but you can if you want to.

emjbishop replied:
I didn't know that, so thank you for telling me! I think I'll keep it for readability for myself for now.

sckott reviewed (commented):
some more comments, looking gr3eat

emjbishop commented on R/build_lookup.R:
```diff
@@ -65,20 +90,46 @@ pad_single_digit <- function(gene_str) {
 #' @return Nothing.
```
I think you're right, I've fixed this in cc44e69d494babaa1832fa11767a03331e1cca1c. For now, `data_dir` is the same as the input folder, but this should be changed using `rappdirs`. I filed an issue to make this change (#9).

emjbishop commented:
> I worry about so much information contained in code comments. Comments are often forgotten about and tend to get out of date. I'd suggest to move important comments within function code into the roxygen2 lines above the function using appropriate roxygen2 tags (e.g., details, section, etc.)

Regarding moving information from comments to function descriptions, I made a note to address this as part of issue #2

emjbishop commented:
Major changes:
- Add a test suite.
- Add GitHub actions to run the test suite with `covr` and run `R CMD check`.

Bug fixes
- Ensure Adaptive lookup tables are made properly by `build_lookup_from_fastas()`.
- Fixed convert_gene() to preserve input row order after merge().

Minor changes:
- Made code style consistent with `styler`.
- Made minor changes in line with `goodpractice::gp()` (e.g. using vapply over sapply).
- Implemented a `verbose` flag that prints message-level information by default.
- Other changes such as checking input validity, ensuring tests and examples work.
